File: Move_Zeroes.swift
/*
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Example:

Input: [0,1,0,3,12]
Output: [1,3,12,0,0]
Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.

Hint 1
In-place means we should not be allocating any space for extra array. But we are allowed to modify the existing array. However, as a first step, try coming up with a solution that makes use of additional space. For this problem as well, first apply the idea discussed using an additional array and the in-place solution will pop up eventually.

Hint 2
A two-pointer approach could be helpful here. The idea would be to have one pointer for iterating the array and another pointer that just works on the non-zero elements of the array.
*/

/*
Solution 2
1 loop
*/
class Solution {
    func moveZeroes(_ nums: inout [Int]) {
        // non-zero index
        var nzIndex = 0
        
        let n = nums.count
        for i in 0..<n {
            if nums[i] != 0 {
                nums.swapAt(nzIndex, i)
                nzIndex += 1
            }
        }
    }
}

/*
Solution 1
array
2 loop 

Time Complexity: O(n)
Space Complexity: O(1)
*/
class Solution {
    func moveZeroes(_ nums: inout [Int]) {
        // non-zero index
        var nzIndex = 0
        
        let n = nums.count
        for i in 0..<n where nums[i] != 0 {
            nums[nzIndex] = nums[i]
            nzIndex += 1
        }
        
        while nzIndex < n {
            nums[nzIndex] = 0
            nzIndex += 1
        }
    }
}

File: Move_Zeroes.java
/*
Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Note that you must do this in-place without making a copy of the array.



Example 1:

Input: nums = [0,1,0,3,12]
Output: [1,3,12,0,0]
Example 2:

Input: nums = [0]
Output: [0]


Constraints:

1 <= nums.length <= 104
-231 <= nums[i] <= 231 - 1


Follow up: Could you minimize the total number of operations done?
*/

/*
use two pointers
i --- point to the original nums
j --- point to the un-zero nums
 */
public class Solution {
    public void moveZeroes(int[] nums) {
        if(nums.length == 0) return;
        int j = 0;
        for(int i = 0; i < nums.length; ++i){
            if(nums[i] != 0){
                nums[j++] = nums[i];
            }
        }

        //set the end of array as zero
        while(j < nums.length){
            nums[j++] = 0;
        }
    }
}
